pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -381,6 +386,90 @@ func (service *SecurityPolicyService) getStores() (*SecurityPolicyStore, *RuleSt
 }
 
 func (service *SecurityPolicyService) createOrUpdateSecurityPolicy(obj *v1alpha1.SecurityPolicy, createdFor string) error {
+	securityPolicyStore, ruleStore, groupStore, _, _ := service.getStores()
+	nsxSecurityPolicy, nsxGroups, _, err := service.buildSecurityPolicy(obj, createdFor)
+	if err != nil {
+		log.Error(err, "failed to build SecurityPolicy")
+		return err
+	}
+
+	if len(nsxSecurityPolicy.Scope) == 0 {
+		log.Info("SecurityPolicy has empty policy-level appliedTo")
+	}
+	indexScope := common.TagValueScopeSecurityPolicyUID
+
+	existingSecurityPolicy := securityPolicyStore.GetByKey(*nsxSecurityPolicy.Id)
+	isChanged := true
+	if existingSecurityPolicy != nil {
+		isChanged = common.CompareResource(SecurityPolicyPtrToComparable(existingSecurityPolicy), SecurityPolicyPtrToComparable(nsxSecurityPolicy))
+	}
+
+	existingRules := ruleStore.GetByIndex(indexScope, string(obj.UID))
+	finalRules := service.updateRules(existingRules, nsxSecurityPolicy.Rules)
+
+	existingGroups := groupStore.GetByIndex(indexScope, string(obj.UID))
+	finalGroups := service.updateGroups(existingGroups, *nsxGroups)
+
+	var finalSecurityPolicy *model.SecurityPolicy
+	if isChanged {
+		finalSecurityPolicy = nsxSecurityPolicy
+	} else {
+		finalSecurityPolicy = existingSecurityPolicy
+	}
+	finalSecurityPolicy.Rules = finalRules
```

Comment:
There're still many verbose codes, which leads to unnecessary line growth and reduced the readability of the code.
Please try this: abstract L389-L419 and L473-L510 to a new function, e.g.
```
func (service *SecurityPolicyService) getFinalSecurityPolicyWithoutVPCResources(obj *v1alpha1.SecurityPolicy, createdFor string)(*v1alpha1.SecurityPolicy, bool, error) {
	/*
	returns:
		*v1alpha1.SecurityPolicy -> finalSecurityPolicy
		bool -> isChanged
		error -> err
	*/
}
```

Reply:
Thanks for advice. I've done a little bit abstraction of getting final securitypolicy from CR for T1 and VPC.